Zqs 1055 create circuits guide (#26)
zapatacomputing/orquestra-core: docs/guides/circuits.rst

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: unitary = state_prep_circ.to_unitary()
  :end-at: ic(type(unitary))

.. include:: /tutorials/beginner_tutorial.rst
    :start-after: icecream-note:
    :end-before: This will output a text description of the circuit

The output we get from those two ``icecream`` statements should be

.. code-block:: text

  ic| unitary.shape: (8, 8)
  ic| type(unitary): <class 'numpy.ndarray'>

Inverting a circuit
-------------------

In some contexts, you may want to invert a circuit. For instance, if instead of :ref:`building a bell state <creating_basic_circuits>`, you want to perform a bell measurement. Or perhaps your circuit requires some `uncomputation <https://quantum.country/search#building-blocks-quantum-search>`_ that you don't want to build manually. To get the inverse of a circuit, just call the ``.inverse()`` method on it:

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: = state_prep_circ.inverse()

>   :end-at: ic(np.allclose(inverse_unitary, np.eye(8)))

Reviewer: I would change the naming in this example.
`invert_state_prep_circ` -> `inverse_state_prep_circ`/`inverse_state_prep_circ`. "invert" is a verb, doesn't fit here.
`invert_test_circ` -> I would call it sth like `combined_circ`? 
`inverse_unitary` -> this one's confusing. It's not really `inverse_unitary`, it's unitary of combined non-inverse and inverse. So I'd call it `combined_unitary` ?

Author: Ah yeah all good points. This was an artifact from what we initially planned to do with this example so the naming is definitely funky. Will fix!

Author: fixed in af446a1749110a289630084291990551053306a0

.. code-block:: text                                                                     
                                                                                
  q_0: ──■───────────────────■────■───────────────────■───────────────────────────
       ┌─┴─┐┌─────────────┐┌─┴─┐  │                   │                           
  q_1: ┤ X ├┤ Rz($\beta$) ├┤ X ├──┼───────────────────┼────■───────────────────■──
       └───┘└─────────────┘└───┘┌─┴─┐┌─────────────┐┌─┴─┐┌─┴─┐┌─────────────┐┌─┴─┐
  q_2: ─────────────────────────┤ X ├┤ Rz($\beta$) ├┤ X ├┤ X ├┤ Rz($\beta$) ├┤ X ├
                                └───┘└─────────────┘└───┘└───┘└─────────────┘└───┘

We'll build up this circuit one connection at a time, using ``+=`` to append ``Circuit`` objects to the end of our existing ``problem_hamiltonian_circ``

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: problem_hamiltonian_circ = Circuit(
  :end-before: unitary = problem_hamiltonian_circ.to_unitary()

> For now, don't worry about what the ``beta`` variables mean in these circuits. We'll cover that later when we talk about :ref:`symbolic gates <symbolic_gates>`. For now, just notice that we started with an initial circuit and we were able to append other circuits to it using ``+=``

Reviewer: I'll put the first sentence before the circuit. Cause I immediately started worrying where `beta` is defined :D

Author: fixed in af446a1749110a289630084291990551053306a0

Let's get the :ref:`unitary matrix <getting_unitary>` again and see if it makes sense for this circuit

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: unitary = problem_hamiltonian_circ.to_unitary()
  :end-at: ic(type(unitary))

The output should look like this:

.. code-block:: text

  ic| unitary.shape: (8, 8)
  ic| type(unitary): <class 'sympy.matrices.immutable.ImmutableDenseMatrix'>

For more info on why this is a sympy matrix as opposed to a numpy ndarray, please see the section on :ref:`symbolic gates <symbolic_gates>`.

Appending individual gates to circuits is also possible, as we'll see when we build up the mixing circuit for this QAOA problem. The mixing circuit looks like this:

.. code-block:: text

       ┌──────────────┐
  q_0: ┤ Rx($\gamma$) ├
       ├──────────────┤
  q_1: ┤ Rx($\gamma$) ├
       ├──────────────┤
  q_2: ┤ Rx($\gamma$) ├
       └──────────────┘

> To build up the mixing circuit, we ust need to put parametrized RX gate on each of the qubits. We can do that with a for loop:

Reviewer: Might be worth mentioning that we have a utility tool that does exactly that as well :) 
https://github.com/zapatacomputing/orquestra-quantum/blob/f58c2ffdba16cef4f4894dea8521c00de44a9797/src/orquestra/quantum/circuits/_generators.py#L14

Author: fixed in af446a1749110a289630084291990551053306a0

Powers of gates and gate exponentials
-------------------------------------

You can add a power of arbitrary non-symbolic gates to your circuit, as well as get the gate exponential. This won't be used in the creation of our QAOA circuit, but we'll still provide examples here. 

**To raise a gate to a power**, you can use ``.power(exponent)`` with the ``exponent`` argument the power you want to raise the gate to. Here we verify that the Hadamard gate raised to the 2nd power is the same as the identity.

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: ic(H.power(2).matrix == sympy.eye(2))
  :end-at: ic(H.power(2).matrix == sympy.eye(2))

Where the output is ``ic| H.power(2).matrix == sympy.eye(2): True``

**To get the gate exponential** of a specific gate, you can use ``.exp``, shown in this example with ``.matrix`` to view what the matrix looks like. It is also generally a good idea to simplify the matrix when possible.

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: ic(sympy.simplify(H.exp.matrix))
  :end-at: ic(sympy.simplify(H.exp.matrix))

.. _wavefunction_operations:

> Gate operations vs Wave Function Operations

Reviewer: I believe Wavefunction is one word.

```suggestion
Gate operations vs Wavefunction Operations
```

Author: Hmmm not according to [Wikipedia](https://en.wikipedia.org/wiki/Wave_function), [Britannica](https://www.britannica.com/science/wave-function), and this [openstax textbook](https://openstax.org/books/university-physics-volume-3/pages/7-1-wave-functions). Unless this is our own standard to have it all one word, I think best to leave it 😄

Reviewer: Ok, fair point! 
Then at least to be consistent with upper-casing please do "Gate operations vs wave function operations"?

Author: Also fair point! Will change 👍

Author: fixed in 9b6a11e3894cfc2026f441d43179d599a8291a33

-------------------------------------------

> While gate operations perform the familiar gates on the qubits of a circuit, `wavefunction operations <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/circuits/_wavefunction_operations.py>`_ can operate on the wavefunction directly. This is convenient if the operation you want to perform is not easily expressible using gates or is not unitary at all. Currently, the only built-in wavefunction operation in Orquestra is the ``MultiPhaseOperation``, which allows a specific phase (given as an angle theta) to be applied to all 2^N components of the wavefunction for an N-qubit circuit.

Reviewer: I'd add info that you can use `WavefunctionOperation` to quickly prepare your simulation in a given state if you know the amplitude vector of this state.

Author: fixed in af446a1749110a289630084291990551053306a0

Wave function operations can be included in the :ref:`creation of a circuit <creating_circuits>`, :ref:`appended later <appending_circuits>`, and :ref:`inspected <inspecting_circuits>` in the same manner GateOperations can be.

.. attention::
  ``Circuit``\ s containing wave function operations generally cannot be run on backends. Simulators are usually able to.

.. _symbolic_gates:

Symbolic gates
==============

Symbolic gates allow the user to specify parametric gates in terms of parameters that can be set to a specific value/angle later. This allows for a few benefits: 

1. You can create the circuit once and bind the parameters later. Sometimes this saves on computational cost, and it allows for more easily understandable implementations of certain circuits, like the QAOA circuit in this example. 

> 2. It gives you a closed formula for the final state of the circuit. This allows users interested in mathematical description and manipulations of the state easier access than could be obtained by varying parameters and running the same circuit multiple times. However, when doing symbolic simulation of the circuit, the performance of the simulation is substantially worse than if the parameters were bound.

Reviewer: Good place for a shameless plug about `SymbolicSimulator` ;)

Author: fixed in af446a1749110a289630084291990551053306a0

We can choose which parameters to bind in our circuit, so first let's just bind the beta parameters:

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: bound_beta_circ =
  :end-at: ic(bound_beta_circ.free_symbols)

This should output ``ic| bound_beta_circ.free_symbols: [gamma_0, gamma_1, gamma_2]`` as the ``beta`` parameters have been bound, but the gamma parameters have not.

Now let's bind the gamma parameters and check what the :ref:`resulting unitary's <getting_unitary>` type is.

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: bound_all_circ =
  :end-at: ic(type(unitary))

We can see in the output that there are no more free parameters and that when all the parameters are bound, the unitary matrix is again a numpy ndarray, rather than a sympy matrix:

.. code-block:: text

  ic| bound_all_circ.free_symbols: []
  ic| type(unitary): <class 'numpy.ndarray'>

.. attention::
  Please note that ``Circuits`` which still have free parameters are not suited to run on most backends. At this time only Zapata's SymbolicSimulator is able to run them.

Using custom gates
==================

> Custom gates let users define gates that can be re-used multiple times. This is especially useful for common sub-circuits like the ``CNOT, RZ, CNOT`` sequence in our ``problem_hamiltonian_circ``. Lets make a custom gate to implement that

Reviewer: ```suggestion
Our libraries let users define gates that can be re-used multiple times. This is especially useful for common sub-circuits like the ``CNOT, RZ, CNOT`` sequence in our ``problem_hamiltonian_circ``. Lets make a custom gate to implement that
```

This should give us ``ic| new_problem_ham_circ.to_unitary() == problem_hamiltonian_circ.to_unitary(): True`` because the two circuits are equivalent.

Getting custom gates from a circuit
-----------------------------------

To get a list of all the custom gate definitions in a ``Circuit``, you can use the ``collect_custom_gate_definitions()`` method. Let's do that on our ``new_problem_ham_circ`` circuit:

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: ic(new_problem_ham_circ.collect_custom_gate_definitions())
  :end-at: ic(new_problem_ham_circ.collect_custom_gate_definitions())

The output should just show our ``ZZ`` gate:

.. code-block:: text

  [CustomGateDefinition(gate_name='ZZ', matrix=Matrix([
  [1.0*exp(-I*theta/2),                  0,                  0,                   0],
  [                  0, 1.0*exp(I*theta/2),                  0,                   0],
  [                  0,                  0, 1.0*exp(I*theta/2),                   0],
  [                  0,                  0,                  0, 1.0*exp(-I*theta/2)]]), params_ordering=(theta,))]

Decompositions
==============

> Decompostion is the process of representing gate operations in (often simpler) different gates. For instance, a SWAP gate can be represented as a series of 3 CNOT gates. Often decomposition is needed to be able to run an algorithm on real hardware as different quantum computers have different sets of gate operations they can physically perform (called native gates).

Reviewer: ```suggestion
Decompostion is the process of representing gate operations using different (often simpler) gates. For instance, a SWAP gate can be represented as a series of 3 CNOT gates. Often, decomposition is needed to be able to run an algorithm on real hardware as different quantum computers have different sets of gate operations they can physically perform (also called "native gates").
```

  q_2: ┤ Rx($\gamma$) ├
       └──────────────┘

To build up the mixing circuit, we need to put parametrized RX gate on each of the qubits. We can do that with a for loop:

.. literalinclude:: /examples/circuits_guide.py
  :language: python
  :start-at: mixing_circ = Circuit()
  :end-at: ic(mixing_circ)

.. note::

>   ``orquestra-quantum`` comes with a `built-in utility function <https://github.com/zapatacomputing/orquestra-quantum/blob/main/src/orquestra/quantum/circuits/_generators.py#L14=>`_ that puts down a layer of the same single-qubit gate acting on all qubits.

Reviewer: I'd say it's better to put a link to the docs rather than to the source code?
Also, I'm not sure if GitHub permalinks are good thing to use in docs.
But I truly don't know, feel free to ignore :)
